### PR TITLE
service/dap: truncate value for expandable types in hover

### PR DESCRIPTION
**Pull request by suzmue**
On hovers, including the full value for complex types is unnecessary, since the user will likely need to expand the children to inspect the values. Additionally, this can really slow down getting the hover values. This change will allow truncating the value
in hover if the variable has a variables reference.

Updates golang/vscode-go#1435

### CODE REVIEW

**Comment by suzmue** [comment]
Hover over `x.s` (truncates the value of hover):
Screen Shot 2021-06-07 at 6 27 27 PM

Hover over `s` (stays the same as currently):
Screen Shot 2021-06-07 at 6 27 21 PM

File: service/dap/server.go [hunk 1]
```diff
@@ -1979,7 +1982,7 @@ func (s *Server) convertVariableWithOpts(v *proc.Variable, qualifiedNameOrExpr s
 			variablesReference = maybeCreateVariableHandle(v)
 		}
 	}
-	canTruncateValue := showFullValue&opts == 0
+	canTruncateValue := showFullValueAll&opts == 0 && (showFullValueBasic&opts == 0 || variablesReference != 0)
```
[the remark this answers is not in the document]
Comment (polinasok) [reply service/dap/server.go:1985]: noChildren would be even more meaningful than no ref.
And there seems to be the does-have-ref and can-have-ref nuance here.
This is indeed hard to follow.
Also it is confusing that we  don't check if we even need to truncate (i.e. value is indeed to long) before going through all the trouble.

File: service/dap/server.go [hunk 2]
```diff
@@ -2057,8 +2060,11 @@ func (s *Server) onEvaluateRequest(request *dap.EvaluateRequest) {
 			}
 		}
 		var opts convertVariableFlags
-		if ctxt == "variables" || ctxt == "hover" || ctxt == "clipboard" {
-			opts |= showFullValue
+		if ctxt == "clipboard" || ctxt == "variables" {
```
Comment (polinasok) [inline service/dap/server.go:2063]: With clipboard option on, when do we get "variables"? Is this in case other editors don't use clipboard?

File: service/dap/server.go [hunk 3]
```diff
@@ -1832,7 +1832,10 @@ type convertVariableFlags uint8
 
 const (
 	skipRef convertVariableFlags = 1 << iota
-	showFullValue
+	// A variable is considered basic if it cannot be expanded to get additional
+	// values.
+	showFullValueBasic
+	showFullValueAll
```
Comment (polinasok) [inline service/dap/server.go:1838]: how about:
showFullValue and showFullValueIfNoChildren

**Review by polinasok — CHANGES_REQUESTED** [review]

File: service/dap/server.go (hunk 2 above)
Comment (suzmue) [reply service/dap/server.go:2063]: This is to support clients that do not use clipboard context

File: service/dap/server.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (suzmue) [reply service/dap/server.go:1985]: With `canHaveRef` helping decide if the value can be truncated, the default behavior is to only truncate if it has children.

I will simplify this code and add a comment describing this.

File: service/dap/server.go (hunk 3 above)
Comment (suzmue) [reply service/dap/server.go]: Not applicable anymore.

**Review by aarzilli — APPROVED** [review]

File: service/dap/server.go [hunk 4]
```diff
@@ -2061,10 +2061,7 @@ func (s *Server) onEvaluateRequest(request *dap.EvaluateRequest) {
 		}
 		var opts convertVariableFlags
 		if ctxt == "clipboard" || ctxt == "variables" {
```
Comment (polinasok) [inline service/dap/server.go:2063]: Let's add a comment that these contexts are usually used to copy the value, which is why we want it in full.

**Review by polinasok — APPROVED** [review]

File: service/dap/server.go (hunk 4 above)
Comment (derekparker) [reply service/dap/server.go:2063]: I'll wait to merge until this is addressed.

**Review by derekparker — APPROVED** [review]
LGTM
